driver/src/main/com/mongodb/client/GridFSUploadStream.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2015 MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client;
+
+import org.bson.types.ObjectId;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * A GridFS OutputStream for uploading data into GridFS
+ *
+ * Provides the {@code fileId} for the file to be uploaded as well as the {@code write} methods of an {@link OutputStream}
+ *
+ * @since 3.1
+ */
+public interface GridFSUploadStream {
+
+    /**
+     * Gets the {@link ObjectId} for the file to be uploaded
+     *
+     * @return the ObjectId for the file to be uploaded
+     */
+    ObjectId getFileId();
+
+    /**
+     * Gets the {@link OutputStream} for uploading data into GridFS
+     *
+     * @return the OutputStream for uploading data into GridFS
+     */
+    OutputStream getOutputStream();
```

Comment:
The other reason is so I can mirror api's with async as much as possible.

For async I will have to have an interface and hide the underlying stream interface because it will be behind a factory supplying me either, nio2, netty or user defined.